cairo-libs/packages/protobuf/src/tests/oneof.cairo

```diff
@@ -0,0 +1,127 @@
+use protobuf::types::message::{
+    ProtoMessage, ProtoOneof, DecodeContext, EncodeContext, EncodeContextTrait, DecodeContextTrait,
+    ProtoCodecImpl,
+};
+use protobuf::types::tag::{WireType, ProtobufTag};
+use protobuf::primitives::numeric::U64AsProtoMessage;
+use protobuf::primitives::array::ByteArrayAsProtoMessage;
+
+#[derive(Drop, Debug, Default, PartialEq)]
+pub struct MessageWithOneof {
+    oneof: Oneof,
+}
+
+pub impl MessageWithOneofAsProtoMessage of ProtoMessage<MessageWithOneof> {
+    fn encode_raw(self: @MessageWithOneof, ref context: EncodeContext) {
+        context.encode_oneof(self.oneof);
+    }
+
+    fn decode_raw(ref context: DecodeContext) -> Option<MessageWithOneof> {
+        let oneof = context.decode_oneof()?;
+        Option::Some(MessageWithOneof { oneof })
+    }
+
+    fn wire_type() -> WireType {
+        WireType::LengthDelimited
+    }
+}
+
+#[derive(Drop, Debug, Default, PartialEq)]
+pub enum Oneof {
+    #[default]
+    Int: u64,
+    String: ByteArray,
+    Enum: Enum,
+}
+
+pub impl OneofAsProtoOneof of ProtoOneof<Oneof> {
+    fn encode_raw(self: @Oneof, ref context: EncodeContext) -> ProtobufTag {
+        match self {
+            Oneof::Int(n) => {
+                n.encode_raw(ref context);
+                let wire_type = ProtoMessage::<u64>::wire_type();
+                ProtobufTag { field_number: 1, wire_type }
+            },
+            Oneof::String(str) => {
+                str.encode_raw(ref context);
+                let wire_type = ProtoMessage::<ByteArray>::wire_type();
+                ProtobufTag { field_number: 2, wire_type }
+            },
+            Oneof::Enum(e) => {
+                let n: u32 = e.into();
+                n.encode_raw(ref context);
+                let wire_type = ProtoMessage::<u32>::wire_type();
+                ProtobufTag { field_number: 3, wire_type }
+            },
+        }
+    }
+
+    fn decode_raw(ref context: DecodeContext, tag: ProtobufTag) -> Option<Oneof> {
+        if tag.field_number == 1 {
+            if tag.wire_type != ProtoMessage::<u64>::wire_type() {
+                return Option::None;
+            }
+            let value = context.decode_field(1)?;
+            Option::Some(Oneof::Int(value))
+        } else if tag.field_number == 2 {
+            if tag.wire_type != ProtoMessage::<ByteArray>::wire_type() {
+                return Option::None;
+            }
+            let value = context.decode_field(2)?;
+            Option::Some(Oneof::String(value))
+        } else if tag.field_number == 3 {
+            if tag.wire_type != ProtoMessage::<u32>::wire_type() {
+                return Option::None;
+            }
+            let value = context.decode_enum(3)?;
+            Option::Some(Oneof::Enum(value))
+        } else {
+            Option::None
+        }
+    }
```

Comment:
we don't need to check for the wire types. It's checked inside `decode_field`


IMO it should look like

```cairo
        match tag.field_number {
          0 => Option::None,
          1 => context.decode_field(1).map(Oneof::Int),
          2 => context.decode_field(2).map(Oneof::String),
          3 => context.decode_field(3).map(Oneof::Enum),
          4 => Option::None,
        }
```

Reply:
Good call [17b0b71](https://github.com/informalsystems/ibc-starknet/pull/301/commits/17b0b7188ede8b0e1b44d4294f9ae86abc12267b)